pkg/cmd/roachtest/tests/cluster_to_cluster.go

```diff
@@ -450,91 +618,98 @@ func registerClusterToCluster(r registry.Registry) {
 			Timeout:         sp.timeout,
 			RequiresLicense: true,
 			Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
-				setup := setupC2C(ctx, t, c, sp.srcNodes, sp.dstNodes)
-				m := c.NewMonitor(ctx, setup.src.nodes.Merge(setup.dst.nodes))
-				metricSnapper := setup.startStatsCollection(ctx, t, c)
-				if initCmd := sp.workload.sourceInitCmd(setup.src.name, setup.src.nodes); initCmd != "" {
-					t.Status("populating source cluster before replication")
-					initStart := timeutil.Now()
-					c.Run(ctx, setup.workloadNode, initCmd)
-					t.L().Printf("src cluster workload initialization took %s minutes", timeutil.Since(initStart).Minutes())
-				}
+				sp.setupC2C(ctx, t, c)
+				m := c.NewMonitor(ctx, sp.setup.src.nodes.Merge(sp.setup.dst.nodes))
+				hc := NewHealthChecker(t, c, sp.setup.src.nodes.Merge(sp.setup.dst.nodes))
+
+				go func() {
+					// Wrap in a goroutine independent of the monitor so the health checker can continue
+					// past future calls of monitor.wait()
+					if err := hc.Runner(ctx); err != nil {
+						t.Fatal(err)
```

Comment:
one peculiar behavior i can't seem to figure out: if this t.Fatal() is called, the  roachtest process panics and the roachprod cluster is not destroyed. 
```
18:56:22 test_impl.go:344: test failure #1: full stack retained in failure_1.log: (cluster_to_cluster.go:649).1: read tcp 10.142.0.85:46552 -> 35.229.107.180:26257: read: connection reset by peer
panic: t.Fatal() was called

goroutine 2843 [running]:
main.(*testImpl).Fatal(0xc00124c4c0?, {0xc0031e80e0?, 0xc000951c40?, 0xc0008f2658?})
        main/pkg/cmd/roachtest/test_impl.go:292 +0x51
github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerClusterToCluster.func1.1()
        github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:649 +0xae
created by github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests.registerClusterToCluster.func1
        github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tests/cluster_to_cluster.go:645 +0x38a
```

If t.Fatal() is called outside of the goroutine (i.e. in the main closure), the test cleanly shuts down and destroys the roachprod cluster. @renatolabs are you aware of this behavior? I imagine wrapping this in a monitor would fix the issue, but I don't understand why and I don't think I want to use the existing monitor, per the comment in the code. Any suggestions?

Reply:
Yes, I'm aware, although this could be better documented. `t.Fatal()` is implemented as a `panic` -- as a consequence, calling it in a goroutine causes the entire test runner to panic and quit unless it's handled in the goroutine itself. There's not much more we can do here. Calling `go` directly in a roachtest is rarely the right thing to do.

A couple of options:

* use a second monitor
* write errors to a channel and read them back in a test's closure, calling `t.Fatal` there.

Reply:
thanks for the explanation! Didn't realize that parent goroutines can't recover from child goroutine panics. See the second commit which adds a little breadcrumb of documentation on this.